### PR TITLE
SAML fixes

### PR DESCRIPTION
**Pull request by relrod**
This ensures that the firstname/lastname/email/etc attributes that are entered into the IdP fields are returned out of the assertion. If additional information is needed admins will have to either set `GET_ALL_EXTRA_DATA` to get it all or they can set `EXTRA_DATA` to get specific fields.

### CODE REVIEW

File: ansible_base/authentication/authenticator_plugins/saml.py [hunk 1]
```diff
@@ -296,6 +296,20 @@ def extra_data(self, user, backend, response, *args, **kwargs):
...
+        if self.database_instance is not None:
+            excluded_fields = ('IDP_URL', 'IDP_X509_CERT', 'IDP_ENTITY_ID')
+            # We are going to auto-include all of the fields like USER_FIRST_NAME, USER_EMAIL, etc.
+            for field, attr_name in SAMLConfiguration.settings_to_enabled_idps_fields.items():
+                if field in excluded_fields:
+                    continue
+                # The fields we want to extra get embedded into the ENABLED_IDPS field so we need to get them out from there
+                field_name = self.database_instance.configuration.get('ENABLED_IDPS', {}).get(idp_string, {}).get(attr_name, None)
+                if field_name in attrs:
+                    data[field_name] = attrs[field_name]
```
Comment (dmzoneill) [inline ansible_base/authentication/authenticator_plugins/saml.py]:
```
# Return early if there's no database instance
if self.database_instance is None:
    return data

excluded_fields = {'IDP_URL', 'IDP_X509_CERT', 'IDP_ENTITY_ID'}
enabled_idps_config = self.database_instance.configuration.get('ENABLED_IDPS', {}).get(idp_string, {})

# Filter fields to exclude specific ones and process only the remaining
included_fields = {
    attr_name: attrs[enabled_idps_config.get(attr_name)]
    for field, attr_name in SAMLConfiguration.settings_to_enabled_idps_fields.items()
    if field not in excluded_fields and enabled_idps_config.get(attr_name) in attrs
}
data.update(included_fields)
```

Screenshot from 2024-11-24 13-43-48

Comment (john-westcott-iv) [reply ansible_base/authentication/authenticator_plugins/saml.py]: I'm not sure I agree with the code readability on generating the included_fields :) 
But I will adopt some of this for sure, thanks!

**Review by jay-steurer — APPROVED** [review]

File: ansible_base/authentication/authenticator_plugins/saml.py [hunk 2]
```diff
@@ -296,6 +296,24 @@ def extra_data(self, user, backend, response, *args, **kwargs):
         if "Group" in attrs:
             response["Group"] = attrs["Group"]
         data = super().extra_data(user, backend, response, *args, **kwargs)
+
+        # Ideally we would always have a DB instance
+        # But if something was mocked in a test or somehow a db_instance just wasn't past in we don't want to error here
```
Comment (BrennanPaciorek) [inline ansible_base/authentication/authenticator_plugins/saml.py]:
```suggestion
        # But if something was mocked in a test or somehow a db_instance just wasn't passed in we don't want to error here
```

**Review by BrennanPaciorek — COMMENTED** [review]
Doing some testing of the PR now, if I com back with nothing, I'll approve this.

**Review by BrennanPaciorek — APPROVED** [review]
Can confirm that SAML auth no longer breaks via keycloak, and unit tests pass for me locally, LGTM.

**Review by dmzoneill — APPROVED** [review]

**Pull request merged**